Moves/updates reports_spec.rb into system

filipefurtad0 opened the pull request:
#### What? Why?

Continues #7299.

Moves/updates reports_spec.rb into system.

<!-- Explain why this change is needed and the solution you propose.
Provide context for others to understand it. -->

This one needed quite a few changes, summarizing:
- case sensitivity on assertions
- when setting dates `fill_in` seems not to work on system specs; now the helper class `select_datetime_from_datepicker` is used instead
- changed some dates from 2015 to 2021 as this massivelly speeds up the spec -> less time scrolling on the flatpickr calendar

#### What should we test?
<!-- List which features should be tested and how. -->

Green build.

#### Release notes
<!-- Write a one liner description of the change to be included in the release notes.
Every PR is worth mentioning, because you did it for a reason. -->

Moves/updates reports_spec.rb into system.

<!-- Please select one for your PR and delete the other. -->
Changelog Category: Technical changes

mkllnk commented on spec/system/admin/reports_spec.rb:
```diff
@@ -1,4 +1,4 @@
-# frozen_string_literal: true
+    # frozen_string_literal: true
```
What's this about?

filipefurtad0 replied:
Oops. No idea how this got here... Fixing it in the next commit.

mkllnk commented on spec/system/admin/reports_spec.rb:
```diff
@@ -295,8 +295,8 @@
       before do
-        Timecop.travel(Time.zone.local(2013, 4, 25, 14, 0, 0)) { order1.finalize! }
-        Timecop.travel(Time.zone.local(2013, 4, 25, 16, 0, 0)) { order2.finalize! }
+        Timecop.travel(Time.zone.local(2021, 4, 25, 14, 0, 0)) { order1.finalize! }
+        Timecop.travel(Time.zone.local(2021, 4, 25, 16, 0, 0)) { order2.finalize! }
```
So we'll have the same problem in a few years time?

I'm wondering if we can use the current date and just travel to a different time.

filipefurtad0 replied:
> So we'll have the same problem in a few years time?
Yes, good point.

> I'm wondering if we can use the current date and just travel to a different time.
That's a good idea. I'll try to figure out the best way to do that.

mkllnk commented on spec/system/admin/reports_spec.rb:
```diff
@@ -560,12 +560,19 @@
         fill_in 'initial_invoice_number', with: '5'
-        fill_in 'invoice_date', with: '2015-02-12'
-        fill_in 'due_date', with: '2015-03-12'
+        find('#invoice_date').click
+        select_datetime_from_datepicker Time.zone.at(Time.zone.local(2021, 02, 12))
+        # hide the datetimepicker
+        find("body").send_keys(:escape)
```
Shall we introduce a helper for this? Maybe used like this:
```rb
pick_date "#invoice_date", Date.new(2021, 2, 12)

# And / or:
pick_datetime "#orders_created_at_from", DateTime.new(2021, 2, 12, 10)
```

I don't think that we have a time zone in the datepicker to consider.

mkllnk commented:
This needs a rebase.

filipefurtad0 commented:
Ok, rebased and addressed all comments, I think.
Build is green - happy for another review round :+1:

jibees commented on .github/workflows/build.yml:
```diff
@@ -34,7 +34,6 @@ jobs:
           - "spec/controllers"
           - "spec/models"
           - "spec/features/admin/[a-o0-9]*_spec.rb"
-          - "spec/features/admin/[p-z]*_spec.rb"
```
🎉

jibees reviewed (approved):
Aggree with

mkllnk commented on spec/support/features/datepicker_helper.rb:
```diff
@@ -47,5 +47,17 @@ def datepicker_month_and_year
       month.to_s + " " + year.to_s
     end
+
+    def pick_date(calendar_selector, date_selector)
+      find(calendar_selector).click
+      select_datetime_from_datepicker date_selector
+      find("body").send_keys(:escape)
+    end
+
+    def pick_datetime(calendar_selector, datetime_selector)
+      find(calendar_selector).click
+      select_datetime_from_datepicker datetime_selector
+      find("body").send_keys(:escape)
+    end
```
These two are exactly the same except the naming, right? It would be nicer to have only one method then. I don't know how to name it though. :wink: So my best idea is to keep the current naming but remove the double-up in code.
```suggestion
    def pick_date(calendar_selector, date_selector)
      pick_datetime(calendar_selector, date_selector)
    end

    def pick_datetime(calendar_selector, datetime_selector)
      find(calendar_selector).click
      select_datetime_from_datepicker datetime_selector
      find("body").send_keys(:escape)
    end
```

What do you think?

mkllnk commented on spec/system/admin/reports_spec.rb:
```diff
@@ -86,25 +86,25 @@
     end
   end
-
+  
```
Accidental whitespace here! :warning: :wink:

mkllnk commented on spec/system/admin/reports_spec.rb:
```diff
@@ -476,12 +481,14 @@
-        fill_in 'invoice_date', with: '2015-02-12'
-        fill_in 'due_date', with: '2015-03-12'
+        
+        pick_datetime '#invoice_date', Date.new(2021, 2, 12)
+        pick_datetime '#due_date', Date.new(2021, 3, 12)
```
Hm, if you use `pick_datetime` with a Date instead of `pick_date` then we don't actually need both methods. :stuck_out_tongue:

mkllnk reviewed (commented):
I'm risking being annoying and request some tiny little changes. :smile:

filipefurtad0 commented:
> I'm risking being annoying and request some tiny little changes.

By all means @mkllnk !! I deeply appreciate the reviews... it's always helpful and constructive. So, yeah - Thank you! :pray: 

Moving to code review again, as I think I've addressed your review in 8d20174 and I can't seem to get rid of rubocop errors... Not sure how to proceed. Other than that, the build is green.

mkllnk commented on spec/system/admin/reports_spec.rb:
```diff
@@ -272,11 +284,11 @@
       product2.taxons = [taxon]
       variant1.on_hand = 10
-      variant1.update_column(:sku, "sku1")
+      variant1.update_attribute(:sku, "sku1")
```
Regarding the Rubocop violation:
```suggestion
      variant1.update!(sku: "sku1")
```

The update method calls validations and the bang version makes sure that the test case fails if a validation fails for an unforseen reason.

While it's good to call validations here, I think, we can't do the same in production code without other changes. In a lot of cases we avoid the validations intentionally. The real solution can be a lot of work and involves calling fewer validations by default and run additional validations when needed.

mkllnk reviewed (approved):
All the Rubocop violations are not related to your changes. You can ignore them and we should add them to the todo list in master. Maybe I'll quickly do that.

mkllnk commented:
I submitted #8570 to avoid these false reports of style violations.
